File: core-cpp/src/main/c++/javolution/internal/osgi/OSGiImpl.cpp
/*
 * Javolution - Java(TM) Solution for Real-Time and Embedded Systems
 * Copyright (C) 2012 - Javolution (http://javolution.org/)
 * All rights reserved.
 */

#include "javolution/internal/osgi/OSGiImpl.hpp"
#include "javolution/lang/Exception.hpp"
#include "javolution/lang/StringBuilder.hpp"
#include "org/osgi/framework/Constants.hpp"
#include "javolution/log/Logging.hpp"
#include "javolution/internal/osgi/BundleImpl.hpp"
#include "javolution/internal/osgi/BundleContextImpl.hpp"

using namespace std;
using namespace javolution::lang;
using namespace javolution::util;
using namespace org::osgi::framework;
using namespace javolution::log;
using namespace javolution::internal::osgi;

void OSGiImpl_API::start(String symbolicName, BundleActivator activator) {
    try {
        BundleImpl bundle = dynamic_cast<BundleImpl_API*>(getBundle(symbolicName).get());
        if (bundle == Type::Null) { // Creates the bundle.
            bundle = new BundleImpl_API(this, symbolicName, activator);
            _bundles->add(bundle);
        }
        bundle->_context = new BundleContextImpl_API(this, bundle);
        bundle->start();
    } catch (Throwable error) {
        Logging_API::error(L"JAVOLUTION", error);
    }
}

void OSGiImpl_API::stop(String symbolicName) {
    try {
        BundleImpl bundle = dynamic_cast<BundleImpl_API*> (getBundle(symbolicName).get());
        if (bundle == Type::Null) throw Exception_API::newInstance("Cannot find bundle " + symbolicName);
        bundle->stop();
        bundle->_context = Type::Null;
    } catch (Throwable error) {
        Logging_API::error(L"JAVOLUTION", error);
    }
}

void OSGiImpl_API::fireServiceEvent(ServiceEvent serviceEvent) {
    ServiceReferenceImpl serviceReference = dynamic_cast<ServiceReferenceImpl_API*>(serviceEvent->getServiceReference().get());
    String serviceName = serviceReference->_serviceName;
    String filter = StringBuilder_API::newInstance()
            ->append('(')->append(Constants_API::OBJECTCLASS)
            ->append('=')->append(serviceName)->append(')')->toString();
    for (int i = 0; i < _bundles->size(); i++) {
        BundleImpl bundle = dynamic_cast<BundleImpl_API*> (_bundles->get(i).get());
        for (int j = 0; j < bundle->_serviceListenerFilters->size(); j++) {
            if (filter->equals(bundle->_serviceListenerFilters->get(j))) {
               ServiceListener listener = bundle->_serviceListeners->get(j);
               listener->serviceChanged(serviceEvent);
            }
        }
    }
}

Bundle OSGiImpl_API::getBundle(String symbolicName) const {
    for (int i = 0; i < _bundles->size(); i++) {
        Bundle bundle = _bundles->get(i);
        if (bundle->getSymbolicName()->equals(symbolicName))
            return bundle;
    }
    return Type::Null;
}

File: javolution/core-java/src/main/java/javolution/internal/osgi/OSGiImpl.java
/*
 * Javolution - Java(TM) Solution for Real-Time and Embedded Systems
 * Copyright (C) 2012 - Javolution (http://javolution.org/)
 * All rights reserved.
 * 
 * Permission to use, copy, modify, and distribute this software is
 * freely granted, provided that this notice is preserved.
 */
package javolution.internal.osgi;

import java.util.ArrayList;
import javolution.context.LogContext;
import javolution.osgi.OSGi;
import org.osgi.framework.Bundle;
import org.osgi.framework.BundleActivator;
import org.osgi.framework.BundleException;
import org.osgi.framework.Constants;
import org.osgi.framework.ServiceEvent;
import org.osgi.framework.ServiceListener;

/**
 *  <p> This class holds the OSGi framework implementation.</p>
 */
public class OSGiImpl extends OSGi {

    ////////////////////////////////////////////////////////////////////////////
    // Converted from OSGiImpl.hpp (C++)
    //
    // Holds the bundles.
    ArrayList<Bundle> bundles;

    /**
     * Creates a new OSGiImpl instance.
     */
    public OSGiImpl() {
        bundles = new ArrayList<Bundle>();
    }

    @Override
    public void start(String symbolicName, BundleActivator activator) {
        try {
            BundleImpl bundle = (BundleImpl) getBundle(symbolicName);
            if (bundle == null) { // Creates the bundle.
                bundle = new BundleImpl(this, symbolicName, activator);
                bundles.add(bundle);
            }
            bundle.context = new BundleContextImpl(this, bundle);
            bundle.start();
        } catch (Throwable error) {
            LogContext.error(error);
        }
    }

    @Override
    public void stop(String symbolicName) {
        try {
            BundleImpl bundle = (BundleImpl) getBundle(symbolicName);
            if (bundle == null)
                throw new BundleException("Cannot find bundle " + symbolicName);
            bundle.stop();
            bundle.context = null;
        } catch (Throwable error) {
            LogContext.error(error);
        }
    }

    void fireServiceEvent(ServiceEvent serviceEvent) {
        ServiceReferenceImpl<?> serviceReference = (ServiceReferenceImpl<?>) serviceEvent.getServiceReference();
        String serviceName = serviceReference.serviceName;
        String filter = "(" + Constants.OBJECTCLASS + "=" + serviceName + ")";
        for (int i = 0; i < bundles.size(); i++) {
            BundleImpl bundle = (BundleImpl) bundles.get(i);
            for (int j = 0; j < bundle.serviceListenerFilters.size(); j++) {
                if (filter.equals(bundle.serviceListenerFilters.get(j))) {
                    ServiceListener listener = bundle.serviceListeners.get(j);
                    listener.serviceChanged(serviceEvent);
                }
            }
        }
    }

    @Override
    public Bundle getBundle(String symbolicName) {
        for (int i = 0; i < bundles.size(); i++) {
            Bundle bundle = bundles.get(i);
            if (bundle.getSymbolicName().equals(symbolicName))
                return bundle;
        }
        return null;
    }
    // End conversion.
    ///////////////////////////////////////////////////////////////////////////
}
